govc: Add checksum validation to govc import.ova

nikhaild opened the pull request:
## Description

This patch enhances govc import.ova command to validate the checksum of an uploaded file (e.g. .vmdk file in .ova) - as computed by the remote server with the checksum listed in the ova manifest (.mf file in .ova). govc import.ova takes a new optional flag to indicate strict checksum validation.
(This govc enhancement needed a small tweak in nfc/lease.go code to actually return manifest data in GetManifest() API.)

This patch also modifies simulator to compute checksum of uploaded file and return that metadata through HttpNfcLeaseGetManifest() call.

govc ova import test now has a new sub-test to test for ova with valid checksum and ova with invalid checksum.

Closes: #1978

## Type of change

Please mark options that are relevant:

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to
  not work as expected)
- [ ] This change requires a documentation update
- [ ] Build related change

## How Has This Been Tested?

- Ran govc ova import existing tests as well as newly added tests
```
$ make install && govc/test/images/update.sh && govc/test/import.bats
...
 ✓ import.ova
 ✓ import.ova with checksum validation
 ✓ import.ova with iso
 ✓ import.ovf
 ✓ import.ovf -host.ipath
 ✓ import.ovf with name in options
 ✓ import.ovf with import.spec result
 ✓ import.ovf with name as argument
 - import.vmdk (skipped: GOVC_TEST_URL not set)
 ✓ import duplicate dvpg names

10 tests, 0 failures, 1 skipped
```

## Checklist:

- [x] My code follows the `CONTRIBUTION`
  [guidelines](https://github.com/vmware/govmomi/blob/main/CONTRIBUTING.md) of
  this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged

dougm reviewed (dismissed):
lgtm, thanks @nikhail !

nikhaild commented:
> lgtm, thanks @nikhail !

Thanks Doug for the quick review!

I tested `govc ova.import` with an ESXi server (for the good & bad check ova files) and found that manifest validation failed (I was using incorrect lookup "key" to associate the client's info with server's manifest info - worked ok with vcsim since I had coded the simulator impl - it was good to test the functionality with an actual ESXi server):
* ova upload client gets (as output of spec import API) nfc.FileItem.DeviceId `/<VM_NAME>/VirtualIDEController0:0`, e.g. `/good-checksum-vm/VirtualIDEController0:0`,
* and what the lease GetManifest() API returns as key is `/<VM_ID>/VirtualIDEController0:0`, e.g. `/21/VirtualIDEController0:0`
It took a while for me to figure out the mapping of data structs, this is the mapping path that works:
```
nfc.FileItem.DeviceId -> types.HttpNfcLeaseInfo.[]HttpNfcLeaseDeviceUrl.ImportKey -> HttpNfcLeaseDeviceUrl.Key == []HttpNfcLeaseManifestEntry.Key
```
I've updated the lookup logic to use the correct key (as well as fixed the simulator impl to match ESXi server). Also fixed a checksum name capitalization issue (`SHA1` vs `sha1`).

Just pushed an update to the patch, will wait for the CI checks to finish.

dougm commented:
Thanks @nikhaild . We see govc tests fail randomly with "Error: The operation was canceled.", something we've not been able to repro locally. The run container test, flakes sometimes due to dockerhub issues.
